lib/Migration/Version1116Date20220110153658.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OCA\Mail\Migration;
+
+use Closure;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Platforms\PostgreSQL94Platform;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version1116Date20220110153658 extends SimpleMigrationStep {
+
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 * @throws Exception
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->getTable('mail_messages');
+		if (!$table->hasIndex('mail_messages_id_flags')) {
+			$table->addIndex(['mailbox_id', 'flag_important', 'flag_deleted', 'flag_seen'], 'mail_messages_id_flags');
+		}
+		if (!$table->hasIndex('mail_messages_id_flags2')) {
+			$table->addIndex(['mailbox_id', 'flag_deleted', 'flag_flagged'], 'mail_messages_id_flags2');
+		}
+
+		// Postgres doesn't allow for shortened indices, so let's skip the last index.
+		if ($schema->getDatabasePlatform() instanceof PostgreSQL94Platform) {
```

Comment:
open a tech dept ticket. this code is prone to explode with any changes to the implementation of the db abstraction. we had the same issue when IDbConnection (or was it the builder?) was assumed to directly implement the Doctrine classes. But since we added a new layer of abstraction that was never the case -> kaboom.

suggestion for a fix: propose an OCP API change with a new accessor that gives us the db type, e.g. as string or number that you can compare to a constant.

Reply:
API change on the ISchemaWrapper class?

Reply:
https://github.com/nextcloud/server/issues/30877